File: fries-leetcode-training/05-linked-list/06-palindrome-linked-list.cpp
/**
 * LeetCode 234. 回文链表
 * https://leetcode.cn/problems/palindrome-linked-list/
 * 
 * 给你一个单链表的头节点head，请你判断该链表是否为回文链表。
 * 
 * 快慢指针 + 反转
 * 
 * 时间复杂度：O(n)
 * 空间复杂度：O(1)
 */

struct ListNode {
    int val;
    ListNode *next;
    ListNode() : val(0), next(nullptr) {}
    ListNode(int x) : val(x), next(nullptr) {}
    ListNode(int x, ListNode *next) : val(x), next(next) {}
};

class Solution {
public:
    bool isPalindrome(ListNode* head) {
        if (!head || !head->next) {
            return true;
        }
        
        // 找到链表中点
        ListNode* slow = head;
        ListNode* fast = head;
        
        while (fast && fast->next) {
            slow = slow->next;
            fast = fast->next->next;
        }
        
        // 反转后半部分
        ListNode* second_half = reverseList(slow);
        
        // 比较前半部分和反转后的后半部分
        ListNode* first_half = head;
        while (second_half) {
            if (first_half->val != second_half->val) {
                return false;
            }
            first_half = first_half->next;
            second_half = second_half->next;
        }
        
        return true;
    }
    
private:
    ListNode* reverseList(ListNode* head) {
        ListNode* prev = nullptr;
        ListNode* current = head;
        
        while (current) {
            ListNode* next_temp = current->next;
            current->next = prev;
            prev = current;
            current = next_temp;
        }
        
        return prev;
    }
};

// 测试函数
#include <iostream>
void testIsPalindrome() {
    Solution solution;
    
    // 测试用例1: 1->2->2->1 (回文)
    ListNode* head1 = new ListNode(1);
    head1->next = new ListNode(2);
    head1->next->next = new ListNode(2);
    head1->next->next->next = new ListNode(1);
    
    bool result1 = solution.isPalindrome(head1);
    cout << "测试1 (1->2->2->1): " << (result1 ? "True" : "False") << endl;  // 期望: True
    
    // 测试用例2: 1->2 (非回文)
    ListNode* head2 = new ListNode(1);
    head2->next = new ListNode(2);
    
    bool result2 = solution.isPalindrome(head2);
    cout << "测试2 (1->2): " << (result2 ? "True" : "False") << endl;  // 期望: False
}

int main() {
    testIsPalindrome();
    return 0;
}

File: fries-leetcode-training/05-linked-list/06-palindrome-linked-list.py
"""
LeetCode 234. 回文链表
https://leetcode.cn/problems/palindrome-linked-list/

给你一个单链表的头节点head，请你判断该链表是否为回文链表。

快慢指针 + 反转

时间复杂度：O(n)
空间复杂度：O(1)
"""

class ListNode:
    def __init__(self, val=0, next=None):
        self.val = val
        self.next = next


def is_palindrome(head):
    """
    回文链表 - 快慢指针 + 反转法
    
    Args:
        head: 链表头节点
        
    Returns:
        是否为回文链表
    """
    if not head or not head.next:
        return True
    
    # 找到链表中点
    slow = fast = head
    while fast and fast.next:
        slow = slow.next
        fast = fast.next.next
    
    # 反转后半部分
    second_half = reverse_list(slow)
    
    # 比较前半部分和反转后的后半部分
    first_half = head
    while second_half:
        if first_half.val != second_half.val:
            return False
        first_half = first_half.next
        second_half = second_half.next
    
    return True


def reverse_list(head):
    """
    反转链表
    
    Args:
        head: 链表头节点
        
    Returns:
        反转后的链表头节点
    """
    prev = None
    current = head
    
    while current:
        next_temp = current.next
        current.next = prev
        prev = current
        current = next_temp
    
    return prev


def test_is_palindrome():
    """测试函数"""
    # 测试用例1: 1->2->2->1 (回文)
    head1 = ListNode(1)
    head1.next = ListNode(2)
    head1.next.next = ListNode(2)
    head1.next.next.next = ListNode(1)
    
    result1 = is_palindrome(head1)
    print(f"测试1 (1->2->2->1): {result1}")  # 期望: True
    
    # 测试用例2: 1->2 (非回文)
    head2 = ListNode(1)
    head2.next = ListNode(2)
    
    result2 = is_palindrome(head2)
    print(f"测试2 (1->2): {result2}")  # 期望: False


if __name__ == "__main__":
    test_is_palindrome()
